lib/thrift/generator/client.ex

```diff
@@ -0,0 +1,270 @@
+
+          defmodule State do
+            defstruct host: nil, port: nil, tcp_opts: nil, timeout: 5000, sock: nil
+          end
+          use Connection
+
+          def init({host, port, tcp_opts, timeout}) do
+            s = %State{host: host,
+                       port: port,
+                       tcp_opts: tcp_opts,
+                       timeout: timeout,
+                       sock: nil}
+
+            {:connect, :init, s}
+          end
+
+          def start_link(host, port, tcp_opts, timeout \\ 5000) do
+            Connection.start_link(__MODULE__, {host, port, tcp_opts, timeout})
+          end
+
+          def send(conn, data), do: Connection.call(conn, {:send, data})
+
+          def recv(conn, bytes, timeout \\ 5000) do
+            Connection.call(conn, {:recv, bytes, timeout})
+          end
+
+          def close(conn), do: Connection.call(conn, :close)
+
+          def connect(_, %{sock: nil,
+                           host: host,
+                           port: port,
+                           tcp_opts: opts,
+                           timeout: timeout}=s) do
+
+            case :gen_tcp.connect(host, port, [active: false, packet: 4] ++ opts, timeout) do
+              {:ok, sock} ->
+                {:ok, %{s | sock: sock}}
+
+              {:error, _} ->
+                {:backoff, 1000, s}
+            end
+          end
+
+          def disconnect(info, %{sock: sock} = s) do
+            :ok = :gen_tcp.close(sock)
+            case info do
+              {:close, from} ->
+                Connection.reply(from, :ok)
+
+              {:error, :closed} ->
+                :error_logger.format("Connection closed~n", [])
+
+              {:error, reason} ->
+                reason = :inet.format_error(reason)
+                :error_logger.format("Connection error: ~s~n", [reason])
+            end
+            {:connect, :reconnect, %{s | sock: nil}}
+          end
+
+          unquote_splicing(functions)
+
+          def handle_call(_, _, %{sock: nil} = s) do
+            {:reply, {:error, :closed}, s}
+          end
+
+          def handle_call({:send, data}, _, %{sock: sock} = s) do
+            case :gen_tcp.send(sock, data) do
+              :ok ->
+                {:reply, :ok, s}
+              {:error, _} = error ->
+                {:disconnect, error, error, s}
+            end
+          end
+
+          def handle_call({:recv, bytes, timeout}, _, %{sock: sock} = s) do
+            case :gen_tcp.recv(sock, bytes, timeout) do
+              {:ok, _} = ok ->
+                {:reply, ok, s}
+
+              {:error, :timeout} = timeout ->
+                {:reply, timeout, s}
+
+              {:error, _} = error ->
+                {:disconnect, error, error, s}
+            end
+          end
+
+          def handle_call(:close, from, s) do
+            {:disconnect, {:close, from}, s}
+          end
+
+          defp deserialize_message_reply(message, rpc_name, sequence_id, reply_module) do
```

Comment:
This function and its nested cases is pretty complex. I think it would be easier to reason about if it were reorganized using some new private functions (assuming this isn't all inlined for some performance benefit).